File: app/api/routers/snapshots.py
"""FastAPI router for snapshot retrieval endpoints.

Exposes endpoints to fetch snapshot metadata (including ML analysis),
query snapshots by date for a camera, and stream snapshot image files
(including from archived ZIP storage).
"""

from datetime import date

from fastapi import APIRouter, Depends, HTTPException
from fastapi.responses import FileResponse, Response
from loguru import logger

from app.api.deps import get_snapshot_service
from app.application.services.snapshot_service import SnapshotService
from app.core.config import settings
from app.domain.schemas import SnapshotRead, SnapshotWithAnalysis, SnapshotAnalysisRead
from app.infrastructure.archive import read_snapshot_from_archive

router = APIRouter(prefix="/api/snapshots", tags=["snapshots"])


@router.get("/{snapshot_id}", response_model=SnapshotWithAnalysis)
async def get_snapshot(
    snapshot_id: int,
    service: SnapshotService = Depends(get_snapshot_service),
):
    """Retrieve a single snapshot by its identifier, including analysis data.

    \f
    Args:
        snapshot_id: The unique identifier of the snapshot.

    Returns:
        The serialized snapshot record with optional ML analysis.

    Raises:
        HTTPException: 404 if the snapshot does not exist.
    """
    snap = await service._uow.snapshots.get_by_id(snapshot_id)
    if not snap:
        logger.warning(f"Snapshot {snapshot_id} not found")
        raise HTTPException(status_code=404, detail="Snapshot not found")
    snap_read = SnapshotRead.model_validate(snap)
    analyses = await service._uow.snapshot_analyses.get_by_snapshot(snapshot_id)
    analysis = SnapshotAnalysisRead.model_validate(analyses[0]) if analyses else None
    return SnapshotWithAnalysis(
        **snap_read.model_dump(),
        analysis=analysis,
    )


@router.get("/{camera_id}/by-date", response_model=list[SnapshotWithAnalysis])
async def get_camera_snapshots(
    camera_id: int,
    snapshot_date: date,
    service: SnapshotService = Depends(get_snapshot_service),
):
    """List all snapshots for a camera on a given date, with analysis data.

    \f
    Args:
        camera_id: The unique identifier of the camera.
        snapshot_date: The date to query snapshots for.

    Returns:
        A list of serialized snapshot records with ML analysis captured on
        that date.
    """
    snapshots = await service.get_camera_snapshots(camera_id, snapshot_date)
    logger.debug(f"Camera {camera_id} snapshots on {snapshot_date}: {len(snapshots)}")
    snapshot_ids = [s.id for s in snapshots]
    analyses = await service._uow.snapshot_analyses.get_by_camera_and_date(camera_id, snapshot_ids)
    analysis_by_snapshot: dict[int, SnapshotAnalysisRead | None] = {}
    for a in analyses:
        analysis_by_snapshot[a.snapshot_id] = SnapshotAnalysisRead.model_validate(a) if a else None

    result = []
    for s in snapshots:
        snap_read = SnapshotRead.model_validate(s)
        result.append(SnapshotWithAnalysis(
            **snap_read.model_dump(),
            analysis=analysis_by_snapshot.get(s.id),
        ))
    return result


@router.get("/image/{snapshot_id}")
async def get_snapshot_image(
    snapshot_id: int,
    service: SnapshotService = Depends(get_snapshot_service),
):
    """Stream a snapshot's JPEG image file.

    \f
    Args:
        snapshot_id: The unique identifier of the snapshot.

    Returns:
        A FileResponse serving the JPEG image.

    Raises:
        HTTPException: 404 if the snapshot record or image file is missing.
    """
    snap = await service._uow.snapshots.get_by_id(snapshot_id)
    if not snap:
        logger.warning(f"Snapshot image: snapshot {snapshot_id} not found")
        raise HTTPException(status_code=404, detail="Snapshot not found")

    full_path = settings.snapshots_dir / snap.image_path
    if full_path.exists():
        logger.debug(f"Serving snapshot image: {full_path}")
        return FileResponse(str(full_path), media_type="image/jpeg")

    # Try serving from archive if the raw file was rotated away
    if snap.archive_path:
        try:
            data = read_snapshot_from_archive(snap.archive_path)
            logger.debug(f"Serving snapshot {snapshot_id} from archive {snap.archive_path}")
            return Response(content=data, media_type="image/jpeg")
        except FileNotFoundError:
            logger.warning(f"Archive file missing for snapshot {snapshot_id}: {snap.archive_path}")
        except Exception as e:
            logger.exception(f"Failed to read snapshot {snapshot_id} from archive: {e}")

    logger.warning(f"Snapshot image file not found: {full_path}, archive={snap.archive_path}")
    raise HTTPException(status_code=404, detail="Image file not found")

File: internal/api/snapshots.go
package api

import (
	"net/http"
	"path/filepath"
	"time"

	"github.com/go-chi/chi/v5"
	"github.com/rs/zerolog/log"

	"github.com/Lmex89/home-cameras/internal/domain"
	"github.com/Lmex89/home-cameras/internal/repository"
)

// snapshotsRoutes registers the /api/snapshots endpoints (parity with
// the legacy snapshots.py router).
//
// Args:
//
//	r: The chi subrouter.
func (s *Server) snapshotsRoutes(r chi.Router) {
	r.Get("/image/{snapshotID}", s.getSnapshotImage)
	r.Get("/{cameraID}/by-date", s.getCameraSnapshots)
	r.Get("/{snapshotID}", s.getSnapshot)
}

// getSnapshot returns one snapshot with its analysis. GET /api/snapshots/{snapshotID}
func (s *Server) getSnapshot(w http.ResponseWriter, r *http.Request) {
	id, ok := pathInt64(w, r, "snapshotID")
	if !ok {
		return
	}
	snap, err := s.deps.Snaps.GetByID(r.Context(), id)
	if err != nil {
		mapRepoErr(w, err, "Snapshot not found")
		return
	}
	analysesRepo := repository.NewSnapshotAnalysisRepository(s.deps.DB)
	analyses, err := analysesRepo.GetBySnapshot(r.Context(), id)
	if err != nil {
		mapRepoErr(w, err, "")
		return
	}
	var analysis *domain.SnapshotAnalysis
	if len(analyses) > 0 {
		analysis = &analyses[0]
	}
	writeJSON(w, http.StatusOK, domain.SnapshotWithAnalysis{Snapshot: *snap, Analysis: analysis})
}

// getCameraSnapshots lists a camera's snapshots for a date with their
// analyses. GET /api/snapshots/{cameraID}/by-date?snapshot_date=YYYY-MM-DD
func (s *Server) getCameraSnapshots(w http.ResponseWriter, r *http.Request) {
	cameraID, ok := pathInt64(w, r, "cameraID")
	if !ok {
		return
	}
	dateRaw := r.URL.Query().Get("snapshot_date")
	day, err := time.ParseInLocation("2006-01-02", dateRaw, time.Local)
	if err != nil {
		writeError(w, http.StatusBadRequest, "snapshot_date must be YYYY-MM-DD")
		return
	}
	snaps, err := s.deps.SnapshotSvc.GetCameraSnapshots(r.Context(), cameraID, day)
	if err != nil {
		mapRepoErr(w, err, "")
		return
	}
	snapshotIDs := make([]int64, 0, len(snaps))
	for _, sn := range snaps {
		snapshotIDs = append(snapshotIDs, sn.ID)
	}
	analysesRepo := repository.NewSnapshotAnalysisRepository(s.deps.DB)
	analyses, err := analysesRepo.GetByCameraAndDate(r.Context(), snapshotIDs)
	if err != nil {
		mapRepoErr(w, err, "")
		return
	}
	analysisBySnap := map[int64]*domain.SnapshotAnalysis{}
	for i := range analyses {
		analysisBySnap[analyses[i].SnapshotID] = &analyses[i]
	}
	out := make([]domain.SnapshotWithAnalysis, 0, len(snaps))
	for _, sn := range snaps {
		out = append(out, domain.SnapshotWithAnalysis{Snapshot: sn, Analysis: analysisBySnap[sn.ID]})
	}
	writeJSON(w, http.StatusOK, out)
}

// getSnapshotImage streams a snapshot's JPEG, falling back to the ZIP
// archive when the raw file has been rotated away.
// GET /api/snapshots/image/{snapshotID}
func (s *Server) getSnapshotImage(w http.ResponseWriter, r *http.Request) {
	id, ok := pathInt64(w, r, "snapshotID")
	if !ok {
		return
	}
	snap, err := s.deps.Snaps.GetByID(r.Context(), id)
	if err != nil {
		mapRepoErr(w, err, "Snapshot not found")
		return
	}
	fullPath := filepath.Join(s.deps.Cfg.SnapshotsDir(), snap.ImagePath)
	if fileExists(fullPath) {
		http.ServeFile(w, r, fullPath)
		return
	}
	if snap.ArchivePath != nil {
		data, err := s.deps.Archive.ReadSnapshot(*snap.ArchivePath)
		if err == nil {
			log.Debug().Int64("snapshot_id", id).Msg("serving snapshot from archive")
			w.Header().Set("Content-Type", "image/jpeg")
			w.WriteHeader(http.StatusOK)
			w.Write(data)
			return
		}
		log.Warn().Err(err).Int64("snapshot_id", id).Msg("archive read failed")
	}
	writeError(w, http.StatusNotFound, "Image file not found")
}
